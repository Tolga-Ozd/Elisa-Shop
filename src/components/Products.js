import React, { useState, useEffect } from "react";
import axios from 'axios'
import styled from "styled-components";
// import {AiOutlineStar, AiFillStar} from 'react-icons/ai'
import Loading from './Loading'
import {Link} from 'react-router-dom';


function Products() {

  const [loading, setLoading] = useState(false);
  const [selectedCategory, setSelectedCategory] = useState('All');
  const [products, setProducts] = useState([]);

  const fetchData = async () => {
    const response = await axios.get("https://fakestoreapi.com/products?limit=30");
    setProducts(response.data);

  };

  useEffect(() => {
    fetchData();

  }, []);


  useEffect(() => {
    setLoading(true);
    setTimeout(() => {
      setLoading(false);
    }, 5000);
  },[])

  const handleCategoryChange = (event) => {
    setSelectedCategory(event.target.value);
  };

  const filteredProducts = selectedCategory === 'All' 
  ? products 
  : products.filter(product => product.category === selectedCategory);


  return (
    
    <Section>
       

       <div className="header-container">
        <h1 className="heading" id="all-products">All Products</h1>
        <select onChange={handleCategoryChange} value={selectedCategory} className="category-select">
          <option value="All">Choose Category</option>
          {Array.from(new Set(products.map(product => product.category))).map(category => (
            <option key={category} value={category}>{category}</option>
          ))}
        </select>
      </div>
    
      {loading ? ( 
       
       <Loading/>

       ) : (
         
      <div className="box-container">
        {filteredProducts.map((pro) => {
          return (
           
            <div className="box" key={pro.id}>
               <Link to={`/products/${pro.id}`} style={{textDecoration:"none"}}>
              <div className="image">
                <img src={pro.image} alt="" />
              </div>
              <div className="content">
                <h3>{pro.title}</h3>
                <div className="price">${pro.price}</div>
              </div>
              </Link>
              
            </div>
            
            
             
          );
        })}
      </div>
     
      )}
    </Section>
  );
}

export default Products;

const Section = styled.section`
.heading{
    font-size: 4rem;
    text-align: center;
    margin-bottom: 2.5rem;
}
.header-container {
    display: flex;
    justify-content: space-between; 
    align-items: center; 
    padding: 0 2rem; 
    h1.heading {
      font-size: 4rem;
      margin: 0; 
    }
.category-select {
      padding: 0.5rem 1rem;
      font-size: 1.5rem;
      cursor: pointer;
      border-radius: 5px;
     border: 1px solid #ccc; 
    }
}
  .box-container {
    display: flex;
    flex-wrap: wrap;
    margin-top: 2rem;
    gap: 2.5rem;
    justify-content: center;
    cursor: pointer;
    .box {
      width: 30rem;
      border-radius: 0.5rem;
      overflow: hidden;
      position: relative;
      box-shadow: 0 .5rem 1rem rgba(0,0,0,.1);
      border: .1rem solid rgba(0,0,0,.3);
      .image{
                height: 35rem;
                padding:3rem;
                width: 100%;
                overflow: hidden;
                img {
                    height: 100%;
                    width: 100%;
                    transition: transform 0.7s ease;
                }
            }
            &:hover .image img{
                transform: scale(1.05);
                opacity: 0.9;
                transition: all 0.7s;
            }
            .content{
                padding:1rem 1.5rem;
                h3{
                    font-size: 2rem;
                    color:black;
                    font-weight: normal;
                }
                .price{
                  
                    padding:.5rem 0;
                    display: flex;
                    gap:.5rem;
                    font-size: 2.2rem;
                    align-items: center;
                }
                .stars{
                    font-size: 2rem;
                    display:flex;
                    svg{
                        font-size: 2rem;
                        color:gold;
                    }
                }
            }
    }
    .category{
        font-size: 2rem;
        text-transform: capitalize;
        background-color:grey;
        text-align: center;
        border-radius: 2rem;
        color:White;
        width:150px;
    }
  }
 
`;