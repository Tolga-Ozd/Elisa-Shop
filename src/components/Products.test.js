import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import axios from 'axios';
import Products from './Products';
import '@testing-library/jest-dom/extend-expect';

jest.mock('axios');

const mockProducts = [
  { id: 1, title: 'Ürün 1', price: '9.99', image: '/image1.jpg', category: 'electronics' },
  { id: 2, title: 'Ürün 2', price: '20.99', image: '/image2.jpg', category: 'jewelery' },
];

describe('Products Component Tests', () => {
  beforeEach(async () => {
    axios.get.mockResolvedValue({ data: mockProducts });
  });

  test('products are fetched and displayed after loading', async () => {
    render(<Products />);
    await waitFor(() => {
      const productElements = screen.getAllByText(/Ürün/);
      expect(productElements).toHaveLength(mockProducts.length);
    });
  });

  test('products are filtered by category', async () => {
    render(<Products />);
    const select = screen.getByRole('combobox');
    fireEvent.change(select, { target: { value: 'electronics' } });

    await waitFor(() => {
      expect(screen.getByText('Ürün 1')).toBeInTheDocument();
    });

    // Ayrı waitFor kullanımı
    await waitFor(() => {
      expect(screen.queryByText('Ürün 2')).not.toBeInTheDocument();
    });
  });

  test('all products are displayed when "All" category is selected', async () => {
    render(<Products />);
    const select = screen.getByRole('combobox');
    fireEvent.change(select, { target: { value: 'All' } });

    // Her beklenen element için ayrı waitFor kullanımı
    await waitFor(() => {
      expect(screen.getByText('Ürün 1')).toBeInTheDocument();
    });

    await waitFor(() => {
      expect(screen.getByText('Ürün 2')).toBeInTheDocument();
    });
  });
});

describe('Products Component Snapshot Test', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(<Products />);
    expect(asFragment()).toMatchSnapshot();
  });
});
