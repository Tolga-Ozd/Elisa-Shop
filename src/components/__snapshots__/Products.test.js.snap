// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Component Snapshot Test should match the snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-gEvEer eLbmEm"
  >
    <div
      class="header-container"
    >
      <h1
        class="heading"
        id="all-products"
      >
        All Products
      </h1>
      <select
        class="category-select"
      >
        <option
          value="All"
        >
          Choose Category
        </option>
      </select>
    </div>
    <section
      class="sc-aXZVg jLVPDW"
    >
      <div />
    </section>
  </section>
</DocumentFragment>
`;
